app/portainer/components/stack-duplication-form/stack-duplication-form-controller.js

```diff
@@ -0,0 +1,78 @@
+angular.module('portainer.app').controller('StackDuplicationFormController', [
+  'Notifications',
+  function StackDuplicationFormController(Notifications) {
+    var ctrl = this;
+
+    ctrl.state = {
+      duplicationInProgress: false,
+      migrationInProgress: false
+    };
+
+    ctrl.formValues = {
+      endpoint: null,
+      newName: ''
+    };
+
+    ctrl.isFormValidForDuplication = isFormValidForDuplication;
+    ctrl.isFormValidForMigration = isFormValidForMigration;
+    ctrl.duplicateStack = duplicateStack;
+    ctrl.migrateStack = migrateStack;
+    ctrl.isMigrationButtonDisabled = isMigrationButtonDisabled;
+
+    function isFormValidForMigration() {
+      return ctrl.formValues.endpoint && ctrl.formValues.endpoint.Id;
+    }
+
+    function isFormValidForDuplication() {
+      return isFormValidForMigration() && ctrl.formValues.newName;
+    }
+
+    function duplicateStack() {
+      if (!ctrl.formValues.newName) {
+        Notifications.error(
+          'Failure',
+          null,
+          'Stack name is required for duplication'
+        );
+        return;
+      }
+      ctrl.state.duplicationInProgress = true;
+      ctrl
```

Comment:
Please have the `onDuplicate` call on the same line here for improved readability.

```
ctrl.onDuplicate({...})
.finally
```